### PR TITLE
[external-assets] Make `observe` support external observables

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

- Update `observe` so that it can handle observable external assets.
- Update `ObserveResult` tests to use `observe`.

## How I Tested These Changes

Updated `ObserveResult` tests.

### CODE REVIEW

**Comment by smackesey** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19898?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #19880** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19880?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #19879** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19879?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #19898** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19898?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #19765** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19765?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #19667** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19667?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19898?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/observe.py [hunk 1]
```diff
@@ -12,22 +15,27 @@
     from ..execution.execute_in_process_result import ExecuteInProcessResult
 
 
+@deprecated_param(
+    param="source_assets", breaking_version="2.0.0", additional_warn_text="Use `assets` instead."
+)
 def observe(
-    source_assets: Sequence[SourceAsset],
+    assets: Optional[Sequence[Union[AssetsDefinition, SourceAsset]]] = None,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/observe.py:22]: why make this optional and default to `None`?

**Review by schrockn — CHANGES_REQUESTED** [review]
Makes sense. Not a big fan of `Optional` and defaulting to `None` in this case. Why change?

File: python_modules/dagster/dagster/_core/definitions/observe.py [hunk 2]
```diff
@@ -12,22 +15,27 @@
     from ..execution.execute_in_process_result import ExecuteInProcessResult
 
 
+@deprecated_param(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/observe.py:18]: This isn't publicly exposed yet, right? If so, we don't need to worry about breaking changes.

File: python_modules/dagster/dagster/_core/definitions/observe.py (hunk 1 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/observe.py]: After Sandy pointed out this isn't public I changed this to just `Sequence[Union[AssetsDefinition, SourceAsset]]`.

File: python_modules/dagster/dagster/_core/definitions/observe.py (hunk 2 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/observe.py]: ah good point I'll remove the deprecation

**Review by schrockn — APPROVED** [review]

**Pull request merged**